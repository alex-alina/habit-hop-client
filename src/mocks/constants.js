export const storageJwtKey = 'jwt';
export const mockJwt =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImRjM2MxMWJmLTU4NGMtNDA5MC1hYThiLTM5ZTQzY2ZhZmNhYiIsImlhdCI6MTY4MTQyMDcwMCwiZXhwIjoxNjgxNDI3OTAwfQ.sYclsZo_DO9zZ2375Kg-AiGZpiVvox1NojPK3eGj9k4';
export const mockUserData = {
  createdAt: '2023-04-25T19:08:15.956Z',
  id: 'dc3c11bf-584c-4090-aa8b-39e43cfafcab',
  email: 'test@example.com',
  firstName: 'Billie',
  lastName: 'Test',
  lastLogin: '2023-04-25T19:08:15.956Z',
};

export const mockGoalFormData = {
  goalDefinition:
    'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...',
  priority: 'main',
  startDate: '2023-03-23',
  endDate: '2023-06-16',
};

export const mockGoalData1 = {
  endDate: '2023-05-06',
  goalDefinition:
    'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...',
  id: 'bec49b7a-7819-453b-9895-347e3982acc5',
  priority: 'main',
  startDate: '2023-04-08',
};

export const mockGoalData2 = {
  endDate: '2023-05-06',
  goalDefinition:
    'I want to learn how to play guitar, and be able to play 3 songs by the end of the year',
  id: '44d96277-99ea-4a83-ab80-deed9148ddc1',
  priority: 'main',
  startDate: '2023-04-08',
};

export const mockGoalData3 = {
  endDate: '2023-05-06',
  goalDefinition:
    'I want to learn how to bake muffins, cupcakes and birthday cakes by baking every weekend for six months',
  id: '1c8a4a14-7fed-4a91-b48b-a9290677043d',
  priority: 'main',
  startDate: '2023-04-08',
};

export const editGoalData = {
  endDate: '2023-05-06',
  goalDefinition:
    'Train for a tripple marathon by running every day for 15km and going to the gym twice per week',
  id: 'bec49b7a-7819-453b-9895-347e3982acc5',
  priority: 'main',
  startDate: '2023-04-08',
};

export const mockHabitFormData = {
  habitDescription: 'Exercise daily',
  habitType: 'develop',
  progressMetric: 'minutes',
};

export const mockHabitData = {
  id: '66c8b782-6238-42a2-ab77-2c03d19b1d0f',
  habitDescription: 'Exercise daily',
  habitType: 'develop',
  progressMetric: 'minutes',
  goal: {
    id: 'bec49b7a-7819-453b-9895-347e3982acc5',
    goalDefinition:
      'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...',
    priority: 'main',
    startDate: '2023-04-08',
    endDate: '2023-05-06',
  },
};

export const mockHabitsData = [
  {
    id: 'aa750fa3-2643-4d84-b4b5-090162ae08b6',
    habitDescription: 'Stop using screens after 20.30',
    habitType: 'break',
    progressMetric: 'is-completed',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
  {
    id: 'cf826e63-c332-48d8-bd50-41d16329fc37',
    habitDescription: 'Do 10 min yoga practice every evening',
    habitType: 'develop',
    progressMetric: 'minutes',
    goal: {
      id: 'bec49b7a-7819-453b-9895-347e3982acc5',
      goalDefinition:
        'Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit...',
      priority: 'secondary',
      startDate: '2023-05-07',
      endDate: '2023-07-15',
    },
  },
];
