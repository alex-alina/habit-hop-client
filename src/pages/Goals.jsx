import React from 'react';
import { Link } from 'react-router-dom';
import Div from '../core-components/Div';

const Home = () => {
  return (
    <Div
      mt={3}
      alignItems="center"
      display="flex"
      flexDirection="column"
      justifyContent="space-between"
      width={300}
      height={100}
    >
      A project without goals is called a hobby
      <Link to="/signup">
        <button>back</button>
      </Link>
      <Link to="/">
        <button>Take me home</button>
      </Link>
    </Div>
  );
};

export default Home;
