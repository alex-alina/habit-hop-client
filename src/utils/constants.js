export const baseUrl = 'http://localhost:3002';

export const localStorageJwtKey = 'userJwt';
