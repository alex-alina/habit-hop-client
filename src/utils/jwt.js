export const jwtData = (jwt) => {
  var base64Url = jwt.split('.')[1];
  var base64 = base64Url.replace('-', '+').replace('_', '/');
  return JSON.parse(window.atob(base64));
};

const currentTime = Date.now() / 1000;

export const extractUserId = (jwt) => jwtData(jwt).id;

export const isExpired = (jwt, now = currentTime) => jwtData(jwt).exp < now;
