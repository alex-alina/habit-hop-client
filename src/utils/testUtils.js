import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../styles/theme';

// export function renderWithTheme(ui, renderOptions) {
//   function Wrapper({ children }) {
//     return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
//   }

//   return {
//     ...render(ui, { wrapper: Wrapper, ...renderOptions }),
//   };
// }

const ThemeProviders = ({ children }) => {
  return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
};

export const renderWithTheme = (ui, options) =>
  render(ui, { wrapper: ThemeProviders, ...options });
