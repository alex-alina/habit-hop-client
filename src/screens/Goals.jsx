import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import {
  addGoal,
  deleteGoal,
  editGoal,
  getGoals,
} from '../actions-reducers/goals';
import { addHabit, getHabits } from '../actions-reducers/habits';
import { logout } from '../actions-reducers/logout';
import { getCurrentUser } from '../actions-reducers/users';
import { ReactComponent as GoalsOverviewImg } from '../assets/illustrations/goals-bg.svg';
import Banner from '../components/Banner';
import FormsOverlay from '../components/FormOverlay';
import GoalCard from '../components/GoalCard/GoalCard';
import IconButton from '../components/IconButton';
import Button from '../core-components/Button';
import Div from '../core-components/Div';
import Header from '../core-components/Heading';
import GoalForm from '../forms/GoalForm';
import HabitForm from '../forms/HabitForm';
import { goalsScreen } from '../text/text';
import { localStorageJwtKey } from '../utils/constants';
import { extractUserId, isExpired } from '../utils/jwt';

const Goals = ({ content = goalsScreen }) => {
  const {
    logoutBtn,
    greeting,
    goalsIntro,
    noGoalsIntro,
    maxNumOfGoalsInfo,
    goalCard,
    addGoalBtn,
    goalsForm,
    habitForm,
  } = content;

  const dispatch = useDispatch();
  const navigate = useNavigate();
  const userToken = localStorage.getItem(localStorageJwtKey);
  const userId = userToken && extractUserId(userToken);

  useEffect(() => {
    if (userId && userToken) {
      dispatch(getCurrentUser({ userId, userToken }));
      dispatch(getGoals({ userId, userToken }));
    }
    if (!userToken || isExpired(userToken)) {
      dispatch(logout());
      navigate('/login');
    }
  }, [userToken]);

  const fullState = useSelector((state) => state);
  const { user, goals } = fullState;

  const stateEntities = Object.keys(fullState);
  const errors = stateEntities
    .map((entity) => {
      const error = fullState[entity].error;
      let message = '';
      if (error && error.message) {
        message = error.message;
      }
      return message;
    })
    .filter((error) => error !== '');

  const goalsList = goals.items;
  const goalsListLen = goalsList && goalsList.length;

  const showAddGoalBtn = goalsListLen > 0 && goalsListLen < 3;
  const hasMaxGoalsNum = goalsListLen === 3;

  const [goalFormIsVisible, setGoalFormVisibility] = useState(false);
  const [isEditMode, setIsEditMode] = useState(false);
  const [editedGoal, setEditedGoal] = useState(null);

  const [habitFormIsVisible, setHabitFormVisibility] = useState(false);
  const [habitGoalId, setHabitGoalId] = useState('');

  const handleCloseOverlay = (formName) => {
    if (formName === 'goal') {
      setGoalFormVisibility(false);
    }

    if (formName === 'habit') {
      setHabitFormVisibility(false);
    }
  };

  const handleAddGoal = (goal) => {
    dispatch(addGoal({ userId, userToken, goal }));
  };

  const handleEditGoal = (updatedGoal, goalId) => {
    dispatch(editGoal({ userId, userToken, goalId, updatedGoal }));
  };

  const handleDeleteGoal = (e, goalId) => {
    e.preventDefault();
    dispatch(deleteGoal({ userId, userToken, goalId }));
  };

  const handleEdit = (e, goal) => {
    e.preventDefault();
    setIsEditMode(true);
    setEditedGoal(goal);
    setGoalFormVisibility(true);
  };

  const handleAddHabit = (habit, goalId) => {
    dispatch(addHabit({ userId, userToken, goalId, habit }));
    setHabitGoalId('');
  };

  const handleHabitFormOverlay = (goalId) => {
    setHabitFormVisibility(true);
    setHabitGoalId(goalId);
  };

  const handleGetHabits = (goalId) => {
    dispatch(getHabits({ userId, userToken, goalId }));
  };

  return (
    <Div width="100%">
      <Div
        zIndex={0}
        width="100%"
        height="100vh"
        display={['none', 'none', 'none', 'flex', 'flex']}
      >
        <GoalsOverviewImg width="100%" height="100%" />
      </Div>

      {goalFormIsVisible && (
        <FormsOverlay closeHandler={() => handleCloseOverlay('goal')}>
          <GoalForm
            content={goalsForm}
            handleSubmit={isEditMode ? handleEditGoal : handleAddGoal}
            handleCloseOverlay={() => handleCloseOverlay('goal')}
            goal={isEditMode ? editedGoal : null}
          />
        </FormsOverlay>
      )}

      {habitFormIsVisible && (
        <FormsOverlay closeHandler={() => handleCloseOverlay('habit')}>
          <HabitForm
            content={habitForm}
            handleCloseOverlay={() => handleCloseOverlay('habit')}
            handleSubmit={handleAddHabit}
            goalId={habitGoalId}
          />
        </FormsOverlay>
      )}

      <Div
        display="flex"
        flexDirection="column"
        alignItems="center"
        alignContent="center"
        width="100%"
        position="absolute"
        top="0"
      >
        <Div
          width="100%"
          display="flex"
          justifyContent="flex-end"
          alignItems="center"
        >
          <Button
            variant="secondaryMd"
            my={3}
            mx={[2, 2, 2, 2, 7]}
            onClick={(e) => {
              e.preventDefault();
              dispatch(logout());
              navigate('/login');
            }}
          >
            {logoutBtn}
          </Button>
        </Div>
        <Div
          display="flex"
          flexDirection="column"
          justifyContent="flex-start"
          alignItems="center"
          mx="auto"
          width={['90%', '90%', '80%', '80%', '65%']}
        >
          {goalsListLen === 0 ? (
            <>
              {user && user.firstName && (
                <Header
                  mt={1}
                  mb={3}
                  textAlign={['center', 'center', 'center', 'left', 'left']}
                >
                  {greeting}
                  {user.firstName},{noGoalsIntro}
                </Header>
              )}

              <GoalForm content={goalsForm} handleSubmit={handleAddGoal} />
            </>
          ) : (
            <Div
              display="flex"
              flexDirection="column"
              alignItems="center"
              mb={6}
              mt={1}
            >
              {user && user.firstName && (
                <Header>
                  {user.firstName}'s {goalsIntro}
                </Header>
              )}
              {errors.map((error, i) => (
                <Banner
                  key={i}
                  color="error"
                  bg="white"
                  fontSize={3}
                  iconName="caution"
                  iconStroke="#922B21"
                  mt={4}
                >
                  {error}
                </Banner>
              ))}

              {hasMaxGoalsNum && (
                <Banner iconName="tips" mt={4}>
                  {maxNumOfGoalsInfo}
                </Banner>
              )}
            </Div>
          )}

          {goalsListLen > 0 &&
            goalsList.map((goal, i) => {
              const goalId = goal.id;
              return (
                <GoalCard
                  key={i}
                  goal={goal}
                  goalCardText={goalCard}
                  handleDelete={(e) => handleDeleteGoal(e, goalId)}
                  handleEdit={(e) => handleEdit(e, goal)}
                  handleHabitFormOverlay={() => handleHabitFormOverlay(goalId)}
                  handleGetHabits={() => handleGetHabits(goalId)}
                />
              );
            })}

          {showAddGoalBtn && (
            <IconButton
              clickHandler={() => {
                setIsEditMode(false);
                setGoalFormVisibility(true);
              }}
              iconName="add-one"
              variant="secondarySm"
              mt={[2, 2, 2, 4, 4]}
              mb={6}
            >
              {addGoalBtn}
            </IconButton>
          )}
        </Div>
      </Div>
    </Div>
  );
};

export default Goals;
