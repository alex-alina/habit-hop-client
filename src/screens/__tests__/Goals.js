import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  editGoalData,
  mockGoalData1,
  mockGoalFormData,
  mockJwt,
} from '../../mocks/constants';
import {
  addGoalException,
  addGoalNetworkError,
  addGoalServerDownError,
  deleteGoalException,
  deleteGoalNetworkError,
  deleteGoalServerDownError,
  editGoalException,
  editGoalNetworkError,
  editGoalServerDownError,
  getEmptyGoalsHandler,
  getGoalsException,
  getGoalsNetworkError,
  getGoalsServerDownError,
  getThreeGoalsHandler,
  getUserException,
  getUserNetworkError,
  getUserServerDownError,
} from '../../mocks/handlers';
import { server } from '../../mocks/server';
import * as jwtModule from '../../utils/jwt';
import { renderWithProvidersAndRouter } from '../../utils/testUtils';
import Goals from '../Goals';

describe('Goals component', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => mockJwt),
        removeItem: jest.fn(() => null),
      },
      writable: true,
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders screen with user data and no goals', async () => {
    server.use(getEmptyGoalsHandler);
    let goalsContainer;

    await act(async () => {
      const { container } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      goalsContainer = container;
    });

    expect(goalsContainer).toBeInTheDocument();
    expect(goalsContainer.innerHTML).toMatchSnapshot();
    expect(screen.getByText(/hi/i)).toBeInTheDocument();
    expect(screen.getByText(/billie/i)).toBeInTheDocument();
    expect(screen.getByText('Set new goal')).toBeInTheDocument();
    expect(screen.getByText(/log out/i)).toBeInTheDocument();
  });

  it('renders screen with error messages: no goals and no user data', async () => {
    server.use(getUserException);
    server.use(getGoalsException);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    expect(screen.getByText('Test Error: User not found')).toBeInTheDocument();
    expect(screen.getByText('Test Error: Goals not found')).toBeInTheDocument();
  });

  it('renders screen with user data and goals error message', async () => {
    server.use(getGoalsException);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    expect(screen.getByText('Test Error: Goals not found')).toBeInTheDocument();
  });

  it('renders screen with no goals, no user data and server down error message', async () => {
    server.use(getUserServerDownError);
    server.use(getGoalsServerDownError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    const error = /There are issues with the server. Please try again later/i;
    expect(screen.getAllByText(error)[0]).toBeInTheDocument();
    expect(screen.getAllByText(error)).toHaveLength(2);
  });

  it('renders screen with network error message', async () => {
    server.use(getUserNetworkError);
    server.use(getGoalsNetworkError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    const error = /Test: some other network error/i;
    expect(screen.getAllByText(error)[0]).toBeInTheDocument();
    expect(screen.getAllByText(error)).toHaveLength(2);
  });

  it('logs out user and redirects to /login', async () => {
    let spy;
    spy = jest.spyOn(jwtModule, 'isExpired');
    spy.mockReturnValue(false);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    expect(spy).toHaveBeenCalled();
    expect(window.location.pathname).toBe('/goals');

    await userEvent.click(screen.getByText(/log out/i));
    expect(window.location.pathname).toBe('/login');
  });

  it('logs out user and redirects to /login when token is expired', async () => {
    let spy;
    spy = jest.spyOn(jwtModule, 'isExpired');
    spy.mockReturnValue(true);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    expect(spy).toHaveBeenCalled();
    expect(window.location.pathname).not.toBe('/goals');
    expect(window.location.pathname).toBe('/login');
  });

  it('displays three goals cards and info banner', async () => {
    server.use(getThreeGoalsHandler);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    const banner = screen.getByText(
      'You can add a maximum of three goals. Edit or replace your goals to fit your current needs.'
    );
    expect(banner).toBeInTheDocument();
    expect(screen.getAllByText(/show habits/i)).toHaveLength(3);
  });

  it('closes the info banner for user with three goals', async () => {
    server.use(getThreeGoalsHandler);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    const banner = screen.queryByText(
      'You can add a maximum of three goals. Edit or replace your goals to fit your current needs.'
    );

    expect(banner).toBeInTheDocument();
    await userEvent.click(
      screen.getByRole('button', { name: /close banner/i })
    );

    expect(banner).not.toBeInTheDocument();
  });

  it('displays the overlay with the goal form when "Add goal" button is clicked', async () => {
    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    expect(
      screen.queryByPlaceholderText(/Go S.M.A.R.T/i)
    ).not.toBeInTheDocument();

    const addGoalBtn = screen.getByText(/add goal/i);
    await userEvent.click(addGoalBtn);
    expect(screen.getByPlaceholderText(/Go S.M.A.R.T/i)).toBeInTheDocument();
  });
});

describe('On the Goals screen, a user can', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => mockJwt),
        removeItem: jest.fn(() => null),
      },
      writable: true,
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  //add goals
  it('add a new goal', async () => {
    server.use(getEmptyGoalsHandler);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      reRender = rerender;
    });

    const goalDescription = screen.getByLabelText('Goal description');
    const goalPriority = screen.getByLabelText('Goal priority');
    const startDate = screen.getByLabelText('Start date');
    const endDate = screen.getByLabelText('End date');
    const addGoalButton = screen.getByText('Set new goal');

    await userEvent.type(goalDescription, mockGoalFormData.goalDefinition);
    await userEvent.type(goalPriority, mockGoalFormData.priority);
    await userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'main' })
    );
    expect(screen.getByRole('option', { name: 'main' }).selected).toBe(true);
    await userEvent.type(startDate, mockGoalFormData.startDate);
    await userEvent.type(endDate, mockGoalFormData.endDate);
    await userEvent.click(addGoalButton);

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    expect(screen.getByText(/overview/i)).toBeInTheDocument();
    expect(screen.getByText('Add goal')).toBeInTheDocument();
    expect(screen.getByText(/show habits/i)).toBeInTheDocument();
    expect(screen.getByText(/billie/i)).toBeInTheDocument();
    expect(
      screen.getByText(mockGoalFormData.goalDefinition)
    ).toBeInTheDocument();
  });

  it('see an error message when goal could not be added', async () => {
    server.use(addGoalException);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });

      reRender = rerender;
    });

    await userEvent.click(screen.getByText('Add goal'));
    const goalDescription = screen.getByLabelText('Goal description');
    const goalPriority = screen.getByLabelText('Goal priority');
    const startDate = screen.getByLabelText('Start date');
    const endDate = screen.getByLabelText('End date');
    const addGoalButton = screen.getByText('Set new goal');

    await userEvent.type(goalDescription, mockGoalFormData.goalDefinition);
    await userEvent.type(goalPriority, mockGoalFormData.priority);
    await userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'main' })
    );
    expect(screen.getByRole('option', { name: 'main' }).selected).toBe(true);
    await userEvent.type(startDate, mockGoalFormData.startDate);
    await userEvent.type(endDate, mockGoalFormData.endDate);
    await userEvent.click(addGoalButton);

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    expect(
      screen.getByText('Test Error: Goal was not added')
    ).toBeInTheDocument();
  });

  it('see a network error message - on add goal', async () => {
    server.use(addGoalNetworkError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getByText('Add goal'));

    await userEvent.type(
      screen.getByLabelText('Goal description'),
      mockGoalFormData.goalDefinition
    );
    await userEvent.type(
      screen.getByLabelText('Goal priority'),
      mockGoalFormData.priority
    );
    await userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'main' })
    );
    await userEvent.type(
      screen.getByLabelText('Start date'),
      mockGoalFormData.startDate
    );
    await userEvent.type(
      screen.getByLabelText('End date'),
      mockGoalFormData.endDate
    );
    await userEvent.click(screen.getByText('Set new goal'));

    expect(
      screen.getByText(/Test: some other network error/i)
    ).toBeInTheDocument();
  });

  it('see a server down error message - on add goal', async () => {
    server.use(addGoalServerDownError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getByText('Add goal'));

    await userEvent.type(
      screen.getByLabelText('Goal description'),
      mockGoalFormData.goalDefinition
    );
    await userEvent.type(
      screen.getByLabelText('Goal priority'),
      mockGoalFormData.priority
    );
    await userEvent.selectOptions(
      screen.getByRole('combobox'),
      screen.getByRole('option', { name: 'main' })
    );
    await userEvent.type(
      screen.getByLabelText('Start date'),
      mockGoalFormData.startDate
    );
    await userEvent.type(
      screen.getByLabelText('End date'),
      mockGoalFormData.endDate
    );
    await userEvent.click(screen.getByText('Set new goal'));

    expect(
      screen.getByText(
        /There are issues with the server. Please try again later/i
      )
    ).toBeInTheDocument();
  });

  //delete goals
  it('delete one goal out of the three displayed', async () => {
    server.use(getThreeGoalsHandler);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      reRender = rerender;
    });

    const deleteGoalButtons = screen.getAllByText(/delete/i);
    expect(deleteGoalButtons).toHaveLength(3);

    await userEvent.click(deleteGoalButtons[0]);

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    const deleteButtons = screen.getAllByText(/delete/i);
    expect(deleteButtons).toHaveLength(2);
  });

  it('see an error message when the goal could not be deleted', async () => {
    server.use(deleteGoalException);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      reRender = rerender;
    });

    const deleteGoalButtons = screen.getAllByText(/delete/i);
    await userEvent.click(deleteGoalButtons[0]);

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    expect(
      screen.getByText('Test Error: Goal was not deleted')
    ).toBeInTheDocument();
  });

  it('see a server down error message - on delete goal', async () => {
    server.use(deleteGoalServerDownError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getAllByText(/delete/i)[0]);
    expect(
      screen.getByText(
        /There are issues with the server. Please try again later/i
      )
    ).toBeInTheDocument();
  });

  it('see a network error message - on delete goal', async () => {
    server.use(deleteGoalNetworkError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getAllByText(/delete/i)[0]);
    expect(
      screen.getByText(/Test: some other network error/i)
    ).toBeInTheDocument();
  });

  //edit goal
  it('edit the goal description of the first goal in the list', async () => {
    server.use(getThreeGoalsHandler);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      reRender = rerender;
    });

    expect(screen.getByText(mockGoalData1.goalDefinition)).toBeInTheDocument();

    const editGoalButtons = screen.getAllByText('Edit');
    expect(editGoalButtons).toHaveLength(3);
    expect(
      screen.queryByText(editGoalData.goalDefinition)
    ).not.toBeInTheDocument();

    await userEvent.click(editGoalButtons[0]);
    await userEvent.type(
      screen.getByLabelText('Goal description'),
      editGoalData.goalDefinition
    );
    await userEvent.click(screen.getByText('Set new goal'));

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    expect(screen.getByText(editGoalData.goalDefinition)).toBeInTheDocument();
  });

  it('see an error message when the goal could not be edited', async () => {
    server.use(editGoalException);
    let reRender;

    await act(async () => {
      const { rerender } = renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
      reRender = rerender;
    });

    const editGoalButtons = screen.getAllByText('Edit');

    await userEvent.click(editGoalButtons[0]);
    await userEvent.type(
      screen.getByLabelText('Goal description'),
      editGoalData.goalDefinition
    );
    await userEvent.click(screen.getByText('Set new goal'));

    await act(async () => {
      reRender(<Goals />, {
        route: '/goals',
      });
    });

    expect(
      screen.getByText('Test Error: Goal was not edited')
    ).toBeInTheDocument();
  });

  it('see a network error message - on edit goal', async () => {
    server.use(editGoalNetworkError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getAllByText('Edit')[0]);
    await userEvent.type(
      screen.getByLabelText('Goal description'),
      editGoalData.goalDefinition
    );
    await userEvent.click(screen.getByText('Set new goal'));
    expect(
      screen.getByText(/Test: some other network error/i)
    ).toBeInTheDocument();
  });

  it('see a server down error message - on edit goal', async () => {
    server.use(editGoalServerDownError);

    await act(async () => {
      renderWithProvidersAndRouter(<Goals />, {
        route: '/goals',
      });
    });

    await userEvent.click(screen.getAllByText('Edit')[0]);
    await userEvent.type(
      screen.getByLabelText('Goal description'),
      editGoalData.goalDefinition
    );
    await userEvent.click(screen.getByText('Set new goal'));
    expect(
      screen.getByText(
        /There are issues with the server. Please try again later/i
      )
    ).toBeInTheDocument();
  });
});
