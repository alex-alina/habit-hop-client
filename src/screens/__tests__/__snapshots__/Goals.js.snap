// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Goals component renders screen with user data and no goals 1`] = `"<div width=\\"100%\\" class=\\"sc-gswNZR hHlltY\\"><div width=\\"100%\\" height=\\"100vh\\" display=\\"none,none,none,flex,flex\\" class=\\"sc-gswNZR itdUVl\\"><svg width=\\"100%\\" height=\\"100%\\">goals-bg.svg</svg></div><div display=\\"flex\\" width=\\"100%\\" class=\\"sc-gswNZR fSPEna\\"><div width=\\"100%\\" display=\\"flex\\" class=\\"sc-gswNZR gLMvmS\\"><button class=\\"sc-eDvSVe hJltOn\\">Log out</button></div><div display=\\"flex\\" width=\\"90%,90%,80%,80%,65%\\" class=\\"sc-gswNZR cRdndm\\"><h1 color=\\"heading\\" font-family=\\"heading\\" font-weight=\\"1\\" class=\\"sc-gKPRtg eqkKZz\\">Hi Billie, add your first goal.</h1><form action=\\"#\\" style=\\"display: flex; justify-content: center; flex-direction: column; background: white; max-width: 400px; width: 80%; margin-top: 10px; margin-bottom: 20px; padding: 30px; border-radius: 10px;\\"><div display=\\"flex\\" class=\\"sc-gswNZR cpRDQJ\\"><label color=\\"label\\" font-size=\\"4\\" font-family=\\"body\\" for=\\"goalDefinition\\" class=\\"sc-jrcTuL eQqIKu\\">Goal description</label><textarea color=\\"text\\" font-family=\\"body\\" font-size=\\"3\\" id=\\"goalDefinition\\" name=\\"goalDefinition\\" placeholder=\\"Go S.M.A.R.T\\" height=\\"100\\" class=\\"sc-eDWCr sc-bqWxrE bgvrIn GSsD\\"></textarea><div class=\\"sc-gswNZR\\"></div></div><div display=\\"flex\\" class=\\"sc-gswNZR cpRDQJ\\"><label color=\\"label\\" font-size=\\"4\\" font-family=\\"body\\" for=\\"priority\\" class=\\"sc-jrcTuL eQqIKu\\">Goal priority</label><select color=\\"text\\" font-family=\\"body\\" font-size=\\"3\\" id=\\"priority\\" name=\\"priority\\" class=\\"sc-ipEyDJ sc-csuSiG fXmrjF bmzVlg\\"><option>Choose priority</option><option>main</option><option>secondary</option><option>tertiary</option></select><div class=\\"sc-gswNZR\\"></div></div><div display=\\"flex\\" class=\\"sc-gswNZR cpRDQJ\\"><label color=\\"label\\" font-size=\\"4\\" font-family=\\"body\\" for=\\"startDate\\" class=\\"sc-jrcTuL eQqIKu\\">Start date</label><input type=\\"date\\" color=\\"text\\" font-family=\\"body\\" font-size=\\"3\\" id=\\"startDate\\" name=\\"startDate\\" format=\\"yyyy-mm-dd\\" class=\\"sc-iBYQkv sc-ftTHYK cRxfSw gtfHWc\\" value=\\"\\"><div class=\\"sc-gswNZR\\"></div></div><div display=\\"flex\\" class=\\"sc-gswNZR cpRDQJ\\"><label color=\\"label\\" font-size=\\"4\\" font-family=\\"body\\" for=\\"endDate\\" class=\\"sc-jrcTuL eQqIKu\\">End date</label><input type=\\"date\\" color=\\"text\\" font-family=\\"body\\" font-size=\\"3\\" id=\\"endDate\\" name=\\"endDate\\" format=\\"yyyy-mm-dd\\" class=\\"sc-iBYQkv sc-ftTHYK cRxfSw gtfHWc\\" value=\\"\\"><div class=\\"sc-gswNZR\\"></div></div><button type=\\"submit\\" class=\\"sc-eDvSVe cOfaGf\\">Set new goal</button></form></div></div></div>"`;
